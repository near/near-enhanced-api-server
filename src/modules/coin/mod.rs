use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
