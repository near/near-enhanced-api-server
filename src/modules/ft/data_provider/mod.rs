mod balance;
mod history;
mod metadata;
mod models;

pub(crate) use balance::{get_ft_balance_by_contract, get_ft_balances};
pub(crate) use history::get_ft_history;
pub(crate) use metadata::get_ft_metadata;
