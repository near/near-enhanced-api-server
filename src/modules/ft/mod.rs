use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    // I'll drop these comments, leaving it only because it's easier for the review
    // app.service(
    //     web::resource("/accounts/{account_id}/NEAR")
    //         .route(web::get().to(resources::get_near_balance)),
    // )
    app.service(
        web::resource("/accounts/{account_id}/FT").route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    // .service(
    //     web::resource("/accounts/{account_id}/NEAR/history")
    //         .route(web::get().to(resources::get_near_history)),
    // )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
