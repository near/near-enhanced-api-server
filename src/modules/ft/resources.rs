use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use super::{data_provider, schemas};
use crate::{db_helpers, errors, modules, types};

#[api_v2_operation(tags(FT))]
/// Get user's FT balances
///
/// This endpoint returns all non-zero FT balances of the given `account_id`,
/// at the given `block_timestamp_nanos`/`block_height`.
///
/// **Limitations**
/// * We may accidentally skip the contract if the balances wasn't updated in the last few months,
///   we are still working on the full coverage.
/// * We currently provide up to 100 items.
///   Full-featured pagination will be provided soon.
pub async fn get_ft_balances(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
    // TODO pagination
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::FtBalancesResponse>> {
    let limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balances = data_provider::get_ft_balances(
        &pool_balances.pool,
        &rpc_client,
        &request.account_id.0,
        &block,
        limit,
    )
    .await?;

    Ok(Json(schemas::FtBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(FT))]
/// Get user's FT balance by contract
///
/// This endpoint returns FT balance of the given `account_id`,
/// for the given `contract_account_id` and `block_timestamp_nanos`/`block_height`.
pub async fn get_ft_balance_by_contract(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceByContractRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtBalanceByContractResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native balance, please use `/accounts/{account_id}/balances/NEAR`".to_string(),
        )
        .into());
    }
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balance = data_provider::get_ft_balance_by_contract(
        &rpc_client,
        &block,
        &request.contract_account_id.0,
        &request.account_id.0,
    )
    .await?;

    Ok(Json(schemas::FtBalanceByContractResponse {
        balance,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(FT))]
/// Get user's FT history by contract
///
/// This endpoint returns the history of FT operations
/// for the given `account_id`, `contract_account_id`, `block_timestamp_nanos`/`block_height`.
///
/// **Limitations**
/// We currently provide the history only for the last few months.
/// The history started from genesis will be served soon.
pub async fn get_ft_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::HistoryRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native history, please use `/accounts/{account_id}/balances/NEAR/history`"
                .to_string(),
        )
        .into());
    }
    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_ft_history(
            &pool,
            &pool_balances.pool,
            &rpc_client,
            &request.contract_account_id.0,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(FT))]
/// Get FT metadata
///
/// This endpoint returns the metadata for the given `contract_account_id`, `block_timestamp_nanos`/`block_height`.
pub async fn get_ft_metadata(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::ContractMetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtContractMetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;

    Ok(Json(schemas::FtContractMetadataResponse {
        metadata: data_provider::get_ft_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
