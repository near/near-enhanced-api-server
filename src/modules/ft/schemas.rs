use paperclip::actix::Apiv2Schema;
use validator::Validate;

use crate::types;

// *** Requests ***

#[derive(
    Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema,
)]
pub struct BalanceRequest {
    #[validate(custom = "crate::errors::validate_account_id")]
    pub account_id: types::AccountId,
}

#[derive(
    Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema,
)]
pub struct BalanceByContractRequest {
    #[validate(custom = "crate::errors::validate_account_id")]
    pub account_id: types::AccountId,
    #[validate(custom = "crate::errors::validate_account_id")]
    pub contract_account_id: types::AccountId,
}

#[derive(
    Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema,
)]
pub struct HistoryRequest {
    #[validate(custom = "crate::errors::validate_account_id")]
    pub account_id: types::AccountId,
    #[validate(custom = "crate::errors::validate_account_id")]
    pub contract_account_id: types::AccountId,
}

#[derive(
    Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema,
)]
pub struct ContractMetadataRequest {
    #[validate(custom = "crate::errors::validate_account_id")]
    pub contract_account_id: types::AccountId,
}

// *** Responses ***

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtBalanceByContractResponse {
    pub balance: FtBalance,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtBalancesResponse {
    pub balances: Vec<FtBalance>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtBalance {
    pub amount: types::U128,
    pub contract_account_id: types::AccountId,
    pub metadata: Metadata,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryResponse {
    pub history: Vec<HistoryItem>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadataResponse {
    pub metadata: FtContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

// ---

/// This type describes the history of the operations (NEAR, FT) for the given user.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryItem {
    // TODO PHASE 2 add index here
    // pub index: types::U128,
    pub involved_account_id: Option<types::AccountId>,
    pub delta_balance: types::I128,
    pub balance: types::U128,
    pub cause: String,
    pub status: String,
    pub metadata: Metadata,
    pub block_timestamp_nanos: types::U64,
    // TODO PHASE 2 add this when we have all the data in the same DB. Now we can't join with blocks
    // pub block_height: types::U64,
}

/// This type describes general Metadata info, collecting the most important fields from different standards in the one format.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct Metadata {
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub decimals: u8,
}

/// The type for FT Contract Metadata. Inspired by
/// https://nomicon.io/Standards/Tokens/FungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadata {
    pub spec: String,
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub reference: Option<String>,
    pub reference_hash: Option<String>,
    pub decimals: u8,
}
