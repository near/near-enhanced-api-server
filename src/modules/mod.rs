use crate::{db_helpers, errors, types};

pub(crate) mod coin;
pub(crate) mod nft;

pub(crate) async fn check_account_exists(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    account_id: &near_primitives::types::AccountId,
    block_height: u64,
) -> crate::Result<()> {
    let request = near_jsonrpc_client::methods::query::RpcQueryRequest {
        block_reference: near_primitives::types::BlockId::Height(block_height).into(),
        request: near_primitives::views::QueryRequest::ViewAccount {
            account_id: account_id.clone(),
        },
    };
    for _ in 0..5 {
        match rpc_client.call(&request).await {
            Err(near_jsonrpc_client::errors::JsonRpcError::ServerError(
                near_jsonrpc_client::errors::JsonRpcServerError::HandlerError(
                    near_jsonrpc_client::methods::query::RpcQueryError::UnknownAccount { .. },
                ),
            )) => {
                return Err(errors::ErrorKind::InvalidInput(format!(
                    "account_id {} does not exist at block_height {}",
                    account_id, block_height
                ))
                .into())
            }
            Err(err) => {
                tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence via JSON RPC failed with: {:?}. Re-trying immediatelly", err);
                continue;
            }
            Ok(response) => {
                if let near_jsonrpc_primitives::types::query::QueryResponseKind::ViewAccount(_) =
                    response.kind
                {
                    return Ok(());
                } else {
                    tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence returned invalid response: {:?}. Re-trying immediatelly", response);
                    continue;
                }
            }
        }
    }
    Err(errors::ErrorKind::InternalError(format!(
        "could not check if account_id {} exists after several attemps",
        account_id
    ))
    .into())
}

pub(crate) async fn check_and_get_history_pagination_params(
    pool: &sqlx::Pool<sqlx::Postgres>,
    pagination_params: types::query_params::HistoryPaginationParams,
) -> crate::Result<types::query_params::HistoryPagination> {
    types::query_params::check_limit(pagination_params.limit)?;
    let pagination = types::query_params::Pagination::from(pagination_params);
    // if pagination_params.after_block_height.is_some() && pagination_params.after_timestamp_nanos.is_some() {
    //     return Err(errors::ErrorKind::InvalidInput(
    //         "Both block_height and block_timestamp_nanos found. Please provide only one of values"
    //             .to_string(),
    //     )
    //         .into());
    // }
    // TODO PHASE 2 take the block from pagination_params
    let block = db_helpers::get_last_block(pool).await?;
    Ok(types::query_params::HistoryPagination {
        block_height: block.height,
        block_timestamp: block.timestamp,
        limit: pagination.limit,
    })
}

#[cfg(test)]
mod tests {
    use crate::db_helpers;

    pub(crate) async fn init_db() -> sqlx::Pool<sqlx::Postgres> {
        dotenv::dotenv().ok();
        let db_url = &std::env::var("DATABASE_URL").expect("failed to get database url");

        sqlx::PgPool::connect(db_url)
            .await
            .expect("failed to connect to the database")
    }

    pub(crate) fn init_rpc() -> near_jsonrpc_client::JsonRpcClient {
        dotenv::dotenv().ok();
        let rpc_url = &std::env::var("RPC_URL").expect("failed to get RPC url");
        let connector = near_jsonrpc_client::JsonRpcClient::new_client();
        connector.connect(rpc_url)
    }

    pub(crate) fn get_block() -> db_helpers::Block {
        db_helpers::Block {
            timestamp: 1655571176644255779,
            height: 68000000,
        }
    }
}
