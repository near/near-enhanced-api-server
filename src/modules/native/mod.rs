use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    );
}
