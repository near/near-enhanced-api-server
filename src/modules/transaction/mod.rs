use paperclip::actix::web;

// mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/transaction/{transaction_hash}")
            .route(web::get().to(resources::get_transaction_by_tx_hash)),
    )
    .service(
        web::resource("/transaction/receipt/{receipt_id}")
            .route(web::get().to(resources::get_transaction_by_receipt)),
    )
    .service(
        web::resource("/transactions/{account_id}")
            .route(web::get().to(resources::get_transactions_by_account)),
    )
    .service(
        web::resource("/transactions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_transactions_by_account_on_contract)),
    )
    .service(
        web::resource("/receipts/{transaction_hash}")
            .route(web::get().to(resources::get_receipts_by_tx_hash)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}")
            .route(web::get().to(resources::get_action_receipts_by_account)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_action_receipts_by_account_on_contract)),
    );
}
