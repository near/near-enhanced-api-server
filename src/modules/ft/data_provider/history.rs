use crate::modules::ft;
use crate::{db_helpers, errors, types};

// TODO PHASE 2 pagination by artificial index added to assets__fungible_token_events
// TODO PHASE 2 change RPC call to DB call by adding absolute amount values to assets__fungible_token_events
// TODO PHASE 2 make the decision about separate FT/MT tables or one table. Pagination implementation depends on this
pub(crate) async fn get_ft_history(
    pool: &sqlx::Pool<sqlx::Postgres>,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: &near_primitives::types::AccountId,
    account_id: &near_primitives::types::AccountId,
    pagination: &types::query_params::HistoryPagination,
) -> crate::Result<Vec<ft::schemas::HistoryItem>> {
    // this is temp solution before we make changes to the DB
    let mut last_balance = super::balance::get_ft_amount(
        rpc_client,
        contract_id.clone(),
        account_id.clone(),
        pagination.block_height,
    )
    .await?;
    let metadata = ft::schemas::Metadata::from(
        super::metadata::get_ft_metadata(rpc_client, contract_id.clone(), pagination.block_height)
            .await?,
    );

    let account_id = account_id.to_string();
    let query = r"
        SELECT
            -- blocks.block_height,
            blocks.block_timestamp,
            assets__fungible_token_events.amount::numeric(45, 0),
            assets__fungible_token_events.event_kind::text cause,
            CASE WHEN execution_outcomes.status IN ('SUCCESS_VALUE', 'SUCCESS_RECEIPT_ID') THEN 'SUCCESS'
                ELSE 'FAILURE'
            END status,
            assets__fungible_token_events.token_old_owner_account_id old_owner_id,
            assets__fungible_token_events.token_new_owner_account_id new_owner_id
        FROM assets__fungible_token_events
            JOIN blocks ON assets__fungible_token_events.emitted_at_block_timestamp = blocks.block_timestamp
            JOIN execution_outcomes ON assets__fungible_token_events.emitted_for_receipt_id = execution_outcomes.receipt_id
        WHERE emitted_by_contract_account_id = $1
            AND (token_old_owner_account_id = $2 OR token_new_owner_account_id = $2)
            AND emitted_at_block_timestamp <= $3::numeric(20, 0)
        ORDER BY emitted_at_block_timestamp desc
        LIMIT $4::numeric(20, 0)
    ";
    let history_info = db_helpers::select_retry_or_panic::<super::models::FtHistoryInfo>(
        pool,
        query,
        &[
            contract_id.to_string(),
            account_id.clone(),
            pagination.block_timestamp.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<ft::schemas::HistoryItem> = vec![];
    for db_info in history_info {
        let mut delta: i128 = types::numeric::to_i128(&db_info.amount)?;
        let balance = last_balance;
        // TODO PHASE 2 maybe we want to change assets__fungible_token_events also to affected/involved?
        let involved_account_id = if account_id == db_info.old_owner_id {
            delta = -delta;
            types::account_id::extract_account_id(&db_info.new_owner_id)?
        } else if account_id == db_info.new_owner_id {
            types::account_id::extract_account_id(&db_info.old_owner_id)?
        } else {
            return Err(
                errors::ErrorKind::InternalError(
                    format!("The account {} should be sender or receiver ({}, {}). If you see this, please create the issue",
                            account_id, db_info.old_owner_id, db_info.new_owner_id)).into(),
            );
        };

        if db_info.status == "SUCCESS" {
            // TODO PHASE 2 this strange error will go away after we add absolute amounts to the DB
            if (last_balance as i128) - delta < 0 {
                return Err(errors::ErrorKind::InternalError(format!(
                    "Balance could not be negative: account {}, contract {}",
                    account_id, contract_id
                ))
                .into());
            }
            last_balance = ((last_balance as i128) - delta) as u128;
        }

        result.push(ft::schemas::HistoryItem {
            cause: db_info.cause.clone(),
            involved_account_id: involved_account_id.map(|id| id.into()),
            delta_balance: delta.into(),
            balance: balance.into(),
            metadata: metadata.clone(),
            block_timestamp_nanos: types::numeric::to_u64(&db_info.block_timestamp)?.into(),
            // block_height: types::numeric::to_u64(&db_info.block_height)?.into(),
            status: db_info.status,
        });
    }
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;

    #[tokio::test]
    async fn test_coin_history() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("usn").unwrap();
        let account = near_primitives::types::AccountId::from_str("pushxo.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_ft_history(&pool, &rpc_client, &contract, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_history_with_failed_receipts() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = db_helpers::Block {
            timestamp: 1651062637353692535,
            height: 64408633,
        };
        let contract =
            near_primitives::types::AccountId::from_str("sweat_token_testing.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("intmainreturn0.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_ft_history(&pool, &rpc_client, &contract, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }
}
