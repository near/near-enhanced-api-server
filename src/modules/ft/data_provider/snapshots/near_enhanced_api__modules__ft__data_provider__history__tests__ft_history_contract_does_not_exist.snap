---
source: src/modules/ft/data_provider/history.rs
expression: balance
---
Err(
    Error {
        code: 400,
        message: "Invalid Input: account_id two-idiots-and-a-half.near does not exist at block_height 80500000",
        retriable: false,
    },
)
